src/vm-vcpu-ref/src/x86_64/gdt.rs

```diff
@@ -1,114 +1,254 @@
+        // Flags::set_available(val).set_present(..) and so on. The flags are
+        // defined in Figure 3-8. Segment Descriptor table of the intel manual.
+        // The getters below (like avl, g, l, p) also are for the flags and their offset
+        // is defined in the spec.
+        SegmentDescriptor(
+            ((u64::from(base) & 0xff00_0000u64) << (56 - 24))
+                | ((u64::from(flags) & 0x0000_f0ffu64) << 40)
+                | ((u64::from(limit) & 0x000f_0000u64) << (48 - 16))
+                | ((u64::from(base) & 0x00ff_ffffu64) << 16)
+                | (u64::from(limit) & 0x0000_ffffu64),
+        )
+    }
 
-pub fn get_db(entry: u64) -> u8 {
-    ((entry & 0x0040_0000_0000_0000) >> 54) as u8
-}
+    fn base(&self) -> u64 {
+        (((self.0) & 0xFF00_0000_0000_0000) >> 32)
+            | (((self.0) & 0x0000_00FF_0000_0000) >> 16)
+            | (((self.0) & 0x0000_0000_FFFF_0000) >> 16)
+    }
 
-pub fn get_l(entry: u64) -> u8 {
-    ((entry & 0x0020_0000_0000_0000) >> 53) as u8
-}
+    fn limit(&self) -> u32 {
+        ((((self.0) & 0x000F_0000_0000_0000) >> 32) | ((self.0) & 0x0000_0000_0000_FFFF)) as u32
+    }
 
-pub fn get_avl(entry: u64) -> u8 {
-    ((entry & 0x0010_0000_0000_0000) >> 52) as u8
-}
+    fn g(&self) -> u8 {
+        ((self.0 & 0x0080_0000_0000_0000) >> 55) as u8
+    }
 
-pub fn get_p(entry: u64) -> u8 {
-    ((entry & 0x0000_8000_0000_0000) >> 47) as u8
-}
+    fn db(&self) -> u8 {
+        ((self.0 & 0x0040_0000_0000_0000) >> 54) as u8
+    }
+
+    fn l(&self) -> u8 {
+        ((self.0 & 0x0020_0000_0000_0000) >> 53) as u8
+    }
+
+    fn avl(&self) -> u8 {
+        ((self.0 & 0x0010_0000_0000_0000) >> 52) as u8
+    }
+
+    fn p(&self) -> u8 {
+        ((self.0 & 0x0000_8000_0000_0000) >> 47) as u8
+    }
+
+    fn dpl(&self) -> u8 {
+        ((self.0 & 0x0000_6000_0000_0000) >> 45) as u8
+    }
+
+    fn s(&self) -> u8 {
+        ((self.0 & 0x0000_1000_0000_0000) >> 44) as u8
+    }
+
+    fn segment_type(&self) -> u8 {
+        ((self.0 & 0x0000_0F00_0000_0000) >> 40) as u8
+    }
 
-pub fn get_dpl(entry: u64) -> u8 {
-    ((entry & 0x0000_6000_0000_0000) >> 45) as u8
+    /// Build a `kvm_segment` from a GDT segment descriptor and the
+    /// selector (`table_index`).
+    ///
+    /// # Arguments
+    ///
+    /// * `entry` - The gdt entry.
+    /// * `table_index` - Index of the entry in the gdt table.
+    ///
+    /// # Example - Create the kvm_segment corresponding to the code
+    ///
+    /// ```rust
+    /// use vm_vcpu_ref::x86_64::gdt::SegmentDescriptor;
+    ///
+    /// let code_descriptor = SegmentDescriptor::from(0xa09b, 0, 0xfffff);
+    /// // The code descriptor is found at position 1 in the GDT.
+    /// // More examples are available here: https://wiki.osdev.org/GDT_Tutorial
+    /// let table_index = 1;
+    /// let code_segment = code_descriptor.create_kvm_segment(table_index);
+    /// ```
+    pub fn create_kvm_segment(&self, table_index: u8) -> kvm_segment {
```

Comment:
Where is the aforementioned `entry` argument?

Reply:
The entry is now `self`. I'll update the documentation.